### ISSUE TITLE
Support dynamic client-level configuration so that we can support multiple studies on the same platform

### ISSUE THREAD

**Issue by shankari** [e-mission/e-mission-docs#739]
This is a large and complex feature, and we definitely want to use some variant of agile programming to develop it.

- Initial features:
  - scan QR code
  - download config
  - store it in local storage

- areas that can be configured
   - server URL
   - summary text
   - terms text
   - show autogen option for auth
   - imperial vs. metric
   - trip end notification enabled

- where to store the config
   - we have not yet created the `JOIN` server, and we want to avoid bulking up the webapp unnecessarily
   - we don't have a way to configure the webapp differently yet
   - the config is not secret and can be transparent
we can store it as a file in a repo on GitHub for now

- Steps to implement so we can test:
   - create config file format
   - put it on github
   - change the initial screen to get the study name via QR code scan or link
   - on app start, load config
   - change the code to reference the config

- Future potential features
   - reload the config periodically to apply changes
   - storing the config in the static webapp which will host the `join` page
   - additional configuration options, including the set of tabs defined in `main.html`

**Comment by shankari** [comment]
we will stick this into a branch for now while we figure out how to ensure that the inclusion of this feature does not make it harder for developers aka how devs can continue to work out of the box.

@PatGendre @lgharib @asiripanich @ericafenyo @jf87 @sebastianbarry 
would like to get feedback from the dev community around various options for supporting configuration, but still keeping it easy for devs. Would really like to avoid multiple branches because of the maintenance overhead.

**Comment by shankari** [comment]
Proposed config file:
- Choosing JSON over YAML temporarily for ease of implementation.
- This proposed config has defaults embedded
- we will initially only support english and spanish for the embedded text
   - **Future work**: need to see how to support additional languages

```
{
    "version": 1,
    "server": {
        "url": "https://openpath-stage.nrel.gov/api/",
        "aggregate_call_auth": "user_only"
    },
    "intro": {
        "program_or_study": "study",
        "program_admin_name": "K. Shankari",
        "program_admin_email": "k.shankari@nrel.gov",
        "deployment_partner_name": "National Renewable Energy Laboratory (NREL)",
        "translated_text": {
            "en": {
                "deployment_name": "Open Access Study",
                "summary_line_1": "enables people to track their travel modes and <b>measure</b> their associated energy use and emissions",
                "summary_line_2": "makes <b>aggregated</b> data on mode shares, trip frequencies, and carbon footprints available via a public dashboard",
                "summary_line_3": "serves as a <b>control group</b> while evaluating behavior change of programs.",
                "short_textual_description": "Transportation is the largest contributor to ...",
                "why_we_collect": "NREL can use this information to ....",
            },
            "es": {
....
            },
            "support_autogen_token": true,
        }
    }
    "display_config": {
        "use_imperial": false,
    },
    "profile_controls": {
        "support_upload": false,
        "trip_end_notification": false
    }
}

```

**Comment by shankari** [comment]
Checked in first full config for the staging environment:
https://github.com/e-mission/nrel-openpath-deploy-configs/blob/main/configs/stage.nrel-op.json
(not sure if we will need to use it, but doesn't hurt)

**Comment by shankari** [comment]
Changed the format slightly after filling in the entries for the first few deployers - one of them has a website to contact and the other includes a phone number

```
git diff configs/
diff --git a/configs/stage.nrel-op.json b/configs/stage.nrel-op.json
index d001cea..654bd29 100644
--- a/configs/stage.nrel-op.json
+++ b/configs/stage.nrel-op.json
@@ -6,8 +6,7 @@
     },
     "intro": {
         "program_or_study": "study",
-        "program_admin_name": "K. Shankari",
-        "program_admin_email": "k.shankari@nrel.gov",
+        "program_admin_contact": "K. Shankari (k.shankari@nrel.gov)",
         "deployment_partner_name": "National Renewable Energy Laboratory (NREL)",
         "translated_text": {
             "en": {
```

**Comment by shankari** [comment]
Checked in the example configs for the programs targeted for this summer:

    - Durham
    - NREL commute study
    - Open Access
    - Yellowstone Teton Clean Cities (Wyoming)

**Comment by shankari** [comment]
Next step is to figure out how to generate a link and QR code for the programs and change the client code to use it to read the config.

Let's start with generating the QR code for the programs.
1. we can generate QR codes through online programs but run the risk of having trackers embedded in them
2. we can generate QR codes manually by hacking the previous emTripLog webapp
3. we can create a GitHub page that automatically generates QR codes

Let's start with (3). It is the most work, but not that much more than hacking. And we might even be able to get the `join` page implemented here and punt on creating another container, at least for now.

**Comment by shankari** [comment]
Done! Picked a URL of the form
nrelopenpath://join_study?label=wyoming&source=github

This has several benefits:
- the `join_study` part tells us what the user wants to do (the `action`)
- using a label instead of a full URL allows us to control the studies that are supported
   - only valid studies are those which have an entry in the github repo that we have write permissions to
- adding a `source` for now ensures that we are can potentially switch to a different hosting location later

**Comment by shankari** [comment]
Once we have received the link in the app, we will store:
- the config URL in as local storage
- the ui config also as local storage
   - or should we just store this as a `config/rw-storage` type
   - let's go with the latter for now to be consistent with the other config objects
   - this would be `config/app_ui_config`
- we would then have a config.js factory which would read the config when the ionic platform is ready
  - for the modules that depend on the config, we need to initialize them only after the config has loaded.
  - we can do this by either:
     - using a singleton/lazy-load pattern for the config where all the dependencies call `Config.getConfig().then`
     - using a broadcast
- we used the broadcast approach to communicate between the diary service and the diary controller, but should have really used an async call instead because it made the code fairly clunky
     - but in this case, there are legitimately multiple listeners, and the code before the broadcast is fairly simple, so maybe this is the best approach after all?

**Comment by shankari** [comment]
Per https://stackoverflow.com/questions/15637133/unsafe-link-in-angular
had to add `$compileProvider.aHrefSanitizationWhitelist(/^\s*(http?|https?|emission):/);`
to get the URLs to not be prefixed by `unsafe`

We can now click on the generated URLs and the app launches.

Next step: create the config class

**Comment by shankari** [comment]
Next step: store the config:
- the config URL in as local storage
- the ui config as a `rw_document` type with key = `config/phone_ui_config`, which means that we need to add server support as well

We also need to handle some corner cases with the ui input:
- if the stored label is the same as the new label, ignore
  - this handles the case in which the user re-launches android https://github.com/e-mission/e-mission-docs/issues/536#issuecomment-629745463)
  - we are passed in the same launch parameters (https://github.com/e-mission/e-mission-docs/issues/536#issuecomment-629893889)

**Comment by shankari** [comment]
While storing values, found that `Promise.all` just wouldn't work for a while. Would not call either `then` or `catch` only `finally`. Separate promises worked. Then tried to document what would and would not work, but then it started working.
Storage + retrieval is done.

**Comment by shankari** [comment]
Next step (penultimate heavy lift):
- cache the config as a variable in the config factory
- change all config locations to read from the config
- this is now forcing us to make the decision: how do we signal that we have read the config?
  - broadcast or lazy load + promise?

**Comment by shankari** [comment]
Plumbed the config through to the intro summary. Still doesn't support HTML in the template text.
The fix is to use `ng-bind-html` (https://www.w3schools.com/angular/ng_ng-bind-html.asp) but we are just not going to support HTML for now.

**Comment by shankari** [comment]
#### Study (open access)

| Summary | Intro | Login |
|--------|--------|-------|
| <img width="448" alt="Screen Shot 2022-06-12 at 6 02 11 PM" src="https://user-images.githubusercontent.com/2423263/173261848-618e8ff1-9335-4778-b998-df0de9200677.png"> | <img width="448" alt="Screen Shot 2022-06-12 at 6 02 31 PM" src="https://user-images.githubusercontent.com/2423263/173261873-bbbbcbdf-8fbf-4d5a-8717-24867eb9c5f4.png"> | <img width="448" alt="Screen Shot 2022-06-12 at 6 03 04 PM" src="https://user-images.githubusercontent.com/2423263/173261914-7c7ddd40-6c49-4187-9820-8470e0ccecc7.png"> |

#### Program (Wyoming)

| Summary | Intro | Login |
|--------|--------|-------|
| <img width="448" alt="Screen Shot 2022-06-12 at 6 06 32 PM" src="https://user-images.githubusercontent.com/2423263/173262179-4f4e40bf-5211-47b5-b535-ee74fcb0e62a.png"> | <img width="448" alt="Screen Shot 2022-06-12 at 6 07 03 PM" src="https://user-images.githubusercontent.com/2423263/173262208-e383ca00-5b72-48da-af67-10ebe2286cd2.png"> | <img width="448" alt="Screen Shot 2022-06-12 at 6 07 21 PM" src="https://user-images.githubusercontent.com/2423263/173262230-f35d448c-7da1-4b1e-a729-d82e45b315bf.png"> |

**Comment by shankari** [comment]
Next steps:
- Finish the other configurations, notably the connection config
- Create the initial link/scan screen
- Handle load when intro is done
- test with scan

Unsure how much we want to rewrite the splash screens, although they really need it.

**Comment by shankari** [comment]
Next, fixed the imperial config:

| Imperial = true | Imperial = false | 
|-------|-------|
| <img width="448" alt="Screen Shot 2022-06-12 at 7 30 11 PM" src="https://user-images.githubusercontent.com/2423263/173268933-f6b62368-071a-4a95-91a4-0128fe5dfd92.png"> | <img width="448" alt="Screen Shot 2022-06-12 at 7 31 52 PM" src="https://user-images.githubusercontent.com/2423263/173269067-ab5fd3a9-b11a-4ab9-8cc5-0f14d0a42791.png"> |

**Comment by shankari** [comment]
Next step: connection config. We currently read the connection config from the `www/json/connectionConfig.json` file, which is one of the most modified config files.
However, we haven't seen people use the `google_auth` recently. And OpenPATH will only ever use autogen/token_list.

We have two options now:
1. combine the values in the dynamic config + values in the file
1. skip the values in the file entirely

Going with option (2) for now for simplicity, and since we don't need to support any other authentication methods right now.

```
{
    "connectUrl": "<e-mission server url, including host and port>",
    "aggregate_call_auth": "no_auth",
    "android": {
        "auth": {
            "method": "prompted-auth",
            "prompt": "<Enter the prompt to display - e.g. Enter your unique token>"
        }
    },
    "ios": {
        "auth": {
            "method": "prompted-auth",
            "prompt": "<Enter the prompt to display - e.g. Enter your unique token>"
        }
    }
}
```

Note that we can also remove the code that reads the file and sets the config in `app.js`

**Comment by shankari** [comment]
Created a new `server_conn.js` by moving the server setting code from `app.js` and setting the connection config based on the server settings from the dynamic config.

**Comment by shankari** [comment]
Next step: handle launch when app launched without clicking on screen

This is a bit tricky, and overlaps a bit with the current preferred screen logic.

The overall logic is:
- load the saved config
- if it exists, go to existing preferred screen logic
- if it doesn't, show join screen

again, preferred screen logic is that we show `intro` if the intro is not done, and the diary/label instead (or we could make that a configuration option too!)

**Comment by shankari** [comment]
The obvious fix is to add an `$ionicPlatform.ready().then(function()` callback to `www/js/config/dynamic_config.js`, and then switch states accordingly. But then the startPrefs code also listens to it, and I'm worried about races.

```
    $ionicPlatform.ready().then(function() {
      Logger.log("ionicPlatform.ready() called " + nTimesCalled+" times!");
      nTimesCalled = nTimesCalled + 1;
      startprefs.startWithPrefs();
      Logger.log("startprefs startup done");
    });
```

**Comment by shankari** [comment]
Aha! Now that we are reading the config directly on load, we are running into the timing issues between generation and delivery.

The dynamic config class was loaded first, so it, in turn, loaded the config and broadcast the notification

```
[Log] DEBUG:Found previously stored ui config, returning it (index.html, line 154)
[Log] Broadcasting event UI_CONFIG_READY (index.html, line 154)
[Log] Received UI_CONFIG_READY notification in server_conn.js, setting config:undefined (index.html, line 154)
[Log] connectionConfig = {"url":"https://durham-openpath.nrel.gov/api/","aggregate_call_auth":"user_only"}:undefined (index.html, line 154)
[Log] Not consented in local storage, need to show consent (index.html, line 154, x3)
[Log] DEBUG:changing state to [object Object] (index.html, line 154)
````

Then start prefs kicks in and changes the state. Which loads the intro cntrl, which registers the listener. Oops! Too late!

```
[Log] loading [object Object] (index.html, line 154)
[Log] Registering for the UI_CONFIG_READY notification in intro.js (index.html, line 154)
[Log] Finished changing state from {"url":"/splash","templateUrl":"templates/splash/splash.html","controller":"SplashCtrl","name":"splash"} to {"url":"/intro","templateUrl":"templates/intro/intro.html","controller":"IntroCtrl","name":"root.intro"} (index.html, line 154)
```

So basically, we can miss configuration through timing issues, and there may be additional timing issues lurking.

🚧 Let's switch to promises instead 🚧

**Comment by PatGendre** [comment]
> @PatGendre @lgharib @asiripanich @ericafenyo @jf87 @sebastianbarry would like to get feedback from the dev community around various options for supporting configuration, but still keeping it easy for devs to get started. Would really like to avoid multiple branches because of the maintenance overhead.

@shankari thanks for this nice feature !
Not sure if this is relevant as this may be at another level, but it'd be nice that the mode / purpose lists would be configurable, because currently if we want to update the mode list we have to build and publish the app.

**Comment by PatGendre** [comment]
@shankari If I understand well:
- this issue aims at implemented limited (but simple) customisation features, to be used for using the new NREL openpath app
- the new openpath app will not used the Ionic Channel feature, which would be much more flexible (we don't use it either)

This app config file on a separate github repo is interesting, I have no idea yet if this can be relevant for us but it is a good idea when the follow-up of the la Rochelle project is more clearly defined.

**Comment by shankari** [comment]
@PatGendre yup!

people will scan a QR code to join a study using the NREL app. Scanning the QR code will download the config and apply it. The config is essentially to switch parts of the main app on and off, or to change some limited portions of the text.

I am not sure I will keep the app config file on a repo, might move it to a separate webapp later. But it will work for now... 😄 
As part of the configuration options, we will be putting in more work to make the existing UI modular. Theoretically, that should make it easier to incorporate the modules into other javascript frameworks, and provide a migration path out of angular1 that allows incremental rewrites.

Would be hear what is going on with the La Rochelle/Cozy Cloud project and the FabMob QC project (@lgharib) so we can coordinate better...

**Comment by shankari** [comment]
While testing, found that the connectionURL configuration was actually broken - the URL was set to `nil`

```
self	CommunicationHelper *	0x600002e13180	0x0000600002e13180
NSObject	NSObject	
_mUrl	NSURL *	nil	0x0000000000000000
_mJsonDict	__NSDictionaryM *	0 key/value pairs	0x000060000203f3c0
_mCompletionHandler	void (^)(NSData *, NSURLResponse *, NSError *)	0x00000001040a6970	
_fetcherService	GTMSessionFetcherService *	0x600001f890e0	0x0000600001f890e0
```

**Comment by shankari** [comment]
That fixed it, but we may want to re-think how we set the connection settings. For example, we should only set the connection settings when the user clicks on a URL. Since the connection settings are saved for re-use, unlike the UI settings, they do not need to be set every time the app loads.

**Comment by shankari** [comment]
Checked in a new "dev" environment to support end-to-end testing by connecting to a local server.
https://github.com/e-mission/nrel-openpath-deploy-configs/commit/6749ca15645e62f62f83c95664ff9979c3673042

**Comment by shankari** [comment]
Now that we are done with server side issues, moving back to promises for notification.
This will also fix the issue where we get an error with the units configuration on first install, since the units are not loaded at the time that the config is loaded, only when we get to the diary/list views.

```
[Log] Broadcasting event UI_CONFIG_CHANGED (index.html, line 154)
[Log] Received UI_CONFIG_CHANGED notification in server_conn.js, filling in templates:undefined (index.html, line 154)
[Log] Broadcasting event UI_CONFIG_READY (index.html, line 154)
[Log] Received UI_CONFIG_READY notification in intro.js, filling in templates:undefined (index.html, line 154)
[Log] Registering for the UI_CONFIG_READY notification in imperial.js (index.html, line 154)
```

**Comment by shankari** [comment]
The naive option 

```
    dc.config_ready = function() {
        return new Promise(function(resolve, reject) {
            Logger.log("Received UI_CONFIG_READY notification in dynamic_config, resolving promise");
            $rootScope.on(UI_CONFIG_READY, (event, newConfig) => resolve(newConfig));
            $timeout(reject("Did not load config within "+LOAD_TIMEOUT+" milliseconds"),
                LOAD_TIMEOUT)
        });
    }
```

does not work because we only register for the notification while creating the promise, way after the notification was generated.

```
[Log] Registering for the UI_CONFIG_CHANGED notification in the server connection (cordova.js, line 1413)
[Log] UI_CONFIG: about to call config_ready function in imperial.js (cordova.js, line 1413)
[Log] UI_CONFIG in config_ready function, about to create promise (cordova.js, line 1413)
[Log] Registering for UI_CONFIG_READY notification in dynamic_config inside the promise (cordova.js, line 1413)
```

wait a minute.... it looks like the broadcast is not getting generated....

**Comment by shankari** [comment]
No, it is that after all

```
[Log] Registering for the UI_CONFIG_CHANGED notification in the server connection (index.html, line 154)
[Log] DEBUG:UI_CONFIG: finished loading config on app start (index.html, line 154)
[Log] Broadcasting event UI_CONFIG_READY (index.html, line 154)
[Log] UI_CONFIG: about to call config_ready function in imperial.js (index.html, line 154)
[Log] UI_CONFIG in config_ready function, about to create promise (index.html, line 154)
[Log] Registering for UI_CONFIG_READY notification in dynamic_config inside the promise (index.html, line 154)
[Log] Registering for the UI_CONFIG_READY notification in imperial.js (index.html, line 154)
```

**Comment by shankari** [comment]
so we have two options:
- singleton promises that are created when the controller is created and resolved directly
- maintain a `ready` state that we check and only create a promise otherwise

Let's go with the second since it seems to be closer to the `ionicPlatform` 

```
    ready: function(cb) {
      // run through tasks to complete now that the device is ready
      if (self.isReady) {
        cb();
      } else {
        // the platform isn't ready yet, add it to this array
        // which will be called once the platform is ready
        readyCallbacks.push(cb);
      }
    },
```

**Comment by shankari** [comment]
That worked! Note that the promise is resolved but the notification is never triggered

```
[Log] Registering for the UI_CONFIG_CHANGED notification in the server connection (index.html, line 154)
[Log] DEBUG:UI_CONFIG: finished loading config on app start (index.html, line 154)
[Log] Broadcasting event UI_CONFIG_READY (index.html, line 154)
[Log] UI_CONFIG: about to call configReady function in imperial.js (index.html, line 154)
[Log] UI_CONFIG in configReady function, resolving immediately (index.html, line 154)
[Log] Registering for the UI_CONFIG_READY notification in imperial.js (index.html, line 154)
[Log] UI_CONFIG: Resolved configReady promise in imperial.js, setting config:undefined (index.html, line 154)
```

**Comment by shankari** [comment]
Next steps:
- Finish the logic around the start screen
- Finish the last two customizations of the control screen or just turn them off in master
- Finish the QR code scan for the token
- Ensure that the configs are all consistent

That's it! Ready to push to stores!

**Comment by shankari** [comment]
Going back to the logic around the start screen (https://github.com/e-mission/e-mission-docs/issues/739#issuecomment-1153494658, https://github.com/e-mission/e-mission-docs/issues/739#issuecomment-1153503792) we can either:
- import the startprefs into config
- import config into startprefs

in either case, we can:
- independently go to the scan screen if the config is not present
- read the config along with the consent and the intro_done and implement the appropriate logic

The first option is easiest to implement, but I am concerned that it might have some issues with races. For example:
- we read that we don't have config, so we go to scan screen
- we read that we don't have intro, so we go to intro screen

and now we can't scan any more

The correct answer is to have all the results and then implement the correct logic. Let's see if we can simplify the logic at the same time, although I am apprehensive about messing with it....

**Comment by shankari** [comment]
Ok, this is not as hard as it seems:

```
startprefs.getPendingOnboardingState
```

reads the consent and done status, so it is fairly straightforward to add another check to it.

startprefs is really a mess right now, though and we should really change it completely.
But that is not a refactoring for today.

**Comment by shankari** [comment]
The change was fairly straightforward (https://github.com/e-mission/e-mission-docs/issues/739#issuecomment-1154685753), except that we also needed to move to intro after downloading and saving the config.

Testing done:
- Launched app: got to the join screen
- Clicked on link: got to intro screen
- Re-launched app: got to the intro screen
- Went through intro: got to the label screen
- Re-launched app: got to the label screen

I declare that this is working

**Comment by shankari** [comment]
Before we work on the UI for the join screen, let's finish up a couple of other changes.

1. I don't know that there is much benefit in actually storing the label and the config RW separately. We always save them together and we always read them together. Storing them separately just adds additional complexity around load/save (`Promise.all`) and leads to more possible error conditions - e.g. label is saved but the config is not, config is restored from server but label is not, etc.

Instead, let's store the label as a field in the config. Having an extra label doesn't hurt the usage of the existing labels in the config, and we can then have a unified version.

2. implement the config for the control screens and maybe add one for the survey so we can skip it in dev mode

**Comment by shankari** [comment]
Implemented the config for the other two fields as well
https://github.com/e-mission/e-mission-phone/pull/842/commits/78c97c35c4ca81220fd5b37902d42b835cb993b5

**Comment by shankari** [comment]
Next steps:
- Create the join screen (based on emTripLog)
- Fix the wrapper on the server
- Create website for staging environment
- Send out to stores for review

**Comment by shankari** [comment]
Copying over from em-base worked remarkably well.

| Initial page | popup description | tips |
|-----|-----|-----|
| <img width="448" alt="Screen Shot 2022-06-14 at 5 12 33 PM" src="https://user-images.githubusercontent.com/2423263/173709959-50a95c75-9c7f-4f23-ba44-f805c39903d9.png"> | <img width="448" alt="Screen Shot 2022-06-14 at 5 12 53 PM" src="https://user-images.githubusercontent.com/2423263/173709978-0cdcea20-53e6-496e-84f9-c801a0115985.png"> | <img width="448" alt="Screen Shot 2022-06-14 at 5 13 25 PM" src="https://user-images.githubusercontent.com/2423263/173710023-43e38685-8e64-450a-a0a9-4b4d9fcf1c3c.png"> |

**Comment by shankari** [comment]
Next steps:
- Create website for staging environment
- Send out to stores for review

I am also going to check in the "scan QR code to login" functionality, but will defer the spanish translations until the weekend.

**Comment by shankari** [comment]
Done with all phone changes except for the "scan QR code to login" functionality, which requires testing on a physical device, which means that I can't use it on the devapp.

Going to merge everything else, then pull to the NREL OpenPATH app, test on a physical device, and then submit.

**Comment by shankari** [comment]
Merged the "Scan QR code to login" change to master https://github.com/e-mission/e-mission-phone/pull/843
and to the UI config branch https://github.com/e-mission/e-mission-phone/pull/844

**Comment by shankari** [comment]
i18n-ized the newly added join pages:
https://github.com/e-mission/e-mission-phone/pull/847
and checked in a spanish translation as well

**Comment by shankari** [comment]
| English | Spanish |
|-----|-----|
| <img width="421" alt="Screen Shot 2022-06-21 at 12 29 12 AM" src="https://user-images.githubusercontent.com/2423263/174741688-24387463-60d3-4795-9158-bd2315ea3e59.png"> | <img width="421" alt="Screen Shot 2022-06-21 at 12 29 29 AM" src="https://user-images.githubusercontent.com/2423263/174741753-7d96ad08-a2d0-4c59-8d71-b5cd71962c03.png"> |
| <img width="421" alt="Screen Shot 2022-06-21 at 12 30 25 AM" src="https://user-images.githubusercontent.com/2423263/174741917-0e6e323b-8d6c-4eb2-9898-17dc7b1380b7.png"> | <img width="421" alt="Screen Shot 2022-06-21 at 12 30 40 AM" src="https://user-images.githubusercontent.com/2423263/174741961-f89ec507-4bc1-4622-abcf-1cc1b5f4190b.png"> |
| <img width="421" alt="Screen Shot 2022-06-21 at 12 31 55 AM" src="https://user-images.githubusercontent.com/2423263/174742185-a318b595-f161-4672-8e29-d5615f7c3b31.png"> | <img width="421" alt="Screen Shot 2022-06-21 at 12 32 08 AM" src="https://user-images.githubusercontent.com/2423263/174742232-cdd030d7-24f6-4ef2-9b22-6ee518af89d5.png"> |
| <img width="421" alt="Screen Shot 2022-06-21 at 12 32 49 AM" src="https://user-images.githubusercontent.com/2423263/174742377-d8182f68-8a9a-48c4-843d-904c8c442333.png"> | <img width="421" alt="Screen Shot 2022-06-21 at 12 33 18 AM" src="https://user-images.githubusercontent.com/2423263/174742453-c8f431e8-7ffc-4c8b-b2da-757a163ae343.png"> |
| <img width="421" alt="Screen Shot 2022-06-21 at 12 37 37 AM" src="https://user-images.githubusercontent.com/2423263/174743269-e9201bc2-0c53-4816-8575-487ddcc8b1ed.png"> | <img width="421" alt="Screen Shot 2022-06-21 at 12 37 56 AM" src="https://user-images.githubusercontent.com/2423263/174743325-04104941-989d-4d00-81ca-ff73e75b5568.png"> |

**Comment by PatGendre** [comment]
@shankari I've seen that the openPaths app was published in the Stores recently, great news :-)
Is there a qrcode to be used for anyone to test the app, or maybe not yet ? Thanks

**Comment by shankari** [comment]
@PatGendre Yay! Had been waiting for Google to approve and refreshing the screen multiple times a day only to have them approve during the night US time. Yes, there is an "open access" study that anybody can join to test the app.

Now that we have published the app, we just need to publish the NREL page with the links to the app. Stay tuned...

**Comment by shankari** [comment]
https://open-access-openpath.nrel.gov/join is now live and will be the primary maintenance focus for the foreseeable future

**Comment by shankari** [comment]
This still has a couple of big pending issues. The most important one is the ability to update the config and have the app read the updated config. that also brings up some questions around backwards compat that we are going to have to handle since we know that the app is not updated at the time it is released.

Tracking that in https://github.com/e-mission/e-mission-docs/issues/765
and closing this issue to avoid a too-long and too-complex issue
